### PR TITLE
IDP: allow using cs3 api to authenticate users

### PR DESCRIPTION
**Pull request by butonic**
The IDP now directly authenticates users using the CS3 API instead of LDAP.

### CODE REVIEW

**Comment by mmattel** [comment]
Do we have any documentation relevant changes?

**Comment by butonic** [comment]
@rhafer changed back to ldap backend by default
@mmattel added a few `desc:` for the idp, no relevant changes as we stay with ldap for now. I'll submut a PR to update the docs for a parallel deployment.

File: extensions/idp/pkg/backends/cs3/identifier/cs3.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+	client := cs3gateway.NewGatewayAPIClient(l)
+
+	res, err := client.Authenticate(ctx, &cs3gateway.AuthenticateRequest{
+		Type:         "machine",
+		ClientId:     "username:" + username,
+		ClientSecret: b.machineAuthAPIKey,
+	})
+	if err != nil || res.Status.Code != cs3rpc.Code_CODE_OK {
+		return nil, nil
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/cs3.go:166]: Again the error is not returned. Is there a reason for this behaviour?

**Review by kobergj — CHANGES_REQUESTED** [review]

File: changelog/unreleased/idp-directly-use-cs3-to-authenticate.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Enhancement: Allow authenticating users via CS3
+
+The IDP can now directly authenticates users using the CS3 API by setting `IDP_IDENTITY_MANAGER="cs3"`.
```
[the remark this answers is not in the document]
Comment (mmattel) [reply changelog/unreleased/idp-directly-use-cs3-to-authenticate.md:3]: HA :point_right: this is definitely docs relevant :point_left: 😄 

Please add an info in the env description about the possible values and what the default value is/should be

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+func (u *cs3User) GivenName() string {
+	return ""
+}
+
+func (u *cs3User) Username() string {
+	return u.u.GetUsername()
+}
+
+func (u *cs3User) ID() int64 {
+	return u.u.GetUidNumber()
```
Comment (rhafer) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:47]: Hm, this seems wrong (especially knowing how we treat the uidNumber in the user backend by default). And IIUC the `ID()` method is optional. Could we just leave it out?

**Review by rhafer — CHANGES_REQUESTED** [review]

File: go.mod [hunk 4]
```diff
@@ -48,6 +48,7 @@ require (
 	github.com/onsi/ginkgo v1.16.5
 	github.com/onsi/ginkgo/v2 v2.1.4
 	github.com/onsi/gomega v1.19.0
+	github.com/orcaman/concurrent-map v1.0.0
```
[the remark this answers is not in the document]
Comment (butonic) [reply go.mod:51]: https://github.com/orcaman/concurrent-map#concurrent-map- points to https://github.com/golang/go/issues/21035

File: extensions/idp/pkg/backends/cs3/identifier/cs3.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		Type:         "basic",
+		ClientId:     username,
+		ClientSecret: password,
+	})
+	if err != nil || res.Status.Code != cs3rpc.Code_CODE_OK {
+		return false, nil, nil, nil, nil
+	}
+	res2, err := client.WhoAmI(ctx, &cs3gateway.WhoAmIRequest{
+		Token: res.Token,
+	})
```
[the remark this answers is not in the document]
Comment (butonic) [reply extensions/idp/pkg/backends/cs3/identifier/cs3.go:102]: nope missed that in the response.

File: extensions/idp/pkg/backends/cs3/identifier/user.go (hunk 3 above)
Comment (butonic) [reply extensions/idp/pkg/backends/cs3/identifier/user.go:47]: for now ... ok

File: extensions/idp/pkg/backends/cs3/identifier/cs3.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+const cs3BackendName = "identifier-cs3"
+
+var cs3SpportedScopes = []string{
+	oidc.ScopeProfile,
+	oidc.ScopeEmail,
+	lico.ScopeUniqueUserID,
+	lico.ScopeRawSubject,
+}
+
+type CS3Backend struct {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/cs3.go:32]: As my friend the hound uses to say:
"exported struct `CS3Backend` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/session.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+type cs3Session struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+	u         *cs3user.User
+	when      time.Time
+}
+
+func (s *cs3Session) User() *cs3user.User {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/session.go:37]: As my friend the hound uses to say:
"exported func `User` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	u *cs3user.User
+}
+
+func newCS3User(u *cs3user.User) (*cs3User, error) {
+	return &cs3User{
+		u: u,
+	}, nil
+}
+
+func (u *cs3User) Subject() string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:18]: As my friend the hound uses to say:
"exported func `Subject` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	return u.u.GetId().GetOpaqueId()
+}
+
+func (u *cs3User) Email() string {
+	return u.u.GetMail()
+}
+
+func (u *cs3User) EmailVerified() bool {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:26]: As my friend the hound uses to say:
"exported func `EmailVerified` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	return u.u.GetMailVerified()
+}
+
+func (u *cs3User) Name() string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:30]: As my friend the hound uses to say:
"exported func `Name` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	return u.u.GetDisplayName()
+}
+
+func (u *cs3User) FamilyName() string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:34]: As my friend the hound uses to say:
"exported func `FamilyName` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	return ""
+}
+
+func (u *cs3User) GivenName() string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:38]: As my friend the hound uses to say:
"exported func `GivenName` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	return ""
+}
+
+func (u *cs3User) Username() string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:42]: As my friend the hound uses to say:
"exported func `Username` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	return u.u.GetUsername()
+}
+
+func (u *cs3User) UniqueID() string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:46]: As my friend the hound uses to say:
"exported func `UniqueID` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	return u.u.GetId().GetOpaqueId()
+}
+
+func (u *cs3User) BackendClaims() map[string]interface{} {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:50]: As my friend the hound uses to say:
"exported func `BackendClaims` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	claims := make(map[string]interface{})
+	claims[konnect.IdentifiedUserIDClaim] = u.u.GetId().GetOpaqueId()
+
+	return claims
+}
+
+func (u *cs3User) BackendScopes() []string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:57]: As my friend the hound uses to say:
"exported func `BackendScopes` should have comment or be unexported"

File: extensions/idp/pkg/backends/cs3/identifier/user.go [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+	return nil
+}
+
+func (u *cs3User) RequiredScopes() []string {
```
Comment (kobergj) [inline extensions/idp/pkg/backends/cs3/identifier/user.go:61]: As my friend the hound uses to say:
"exported func `RequiredScopes` should have comment or be unexported"

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>Core-API-Tests-ocis-storage-7</strong>](https://drone.owncloud.com/owncloud/ocis/11978/45/1) failed. Further test are cancelled...

**Review by kobergj — APPROVED** [review]

**Pull request merged**